### DISCUSSION TITLE
Nested fields?

### DISCUSSION THREAD

**Discussion by ConnectedSystems** [CliMA/Oceananigans.jl#3097] [General]
Hello!

I'm having a conversation with a colleague about potentially using Oceananigans instead of OceanParcels.
A key requirement is the need to represented nested fields (see notebook linked below).

https://nbviewer.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_NestedFields.ipynb

Is this possible with Oceananigans? We both had a brief look over the documentation and nothing stuck out for us. Perhaps the terminology is different?

**Comment by Chinenyeani1986** [comment]
To add to @ConnectedSystems question. We are interested in fitting a high-resolution field inside a low-resolution field with 3D velocities.

**Reply by glwagner** [reply]
Can you elaborate on what a "nested field" is?

**Reply by Chinenyeani1986** [reply]
A nested field is a combination of different fields. For example, combining a high-resolution field that covers a small area and a low-resolution field that covers a large area including the area covered by the high-resolution field. This will allow the continuous tracking of particles that leave the domain of the high-resolution field.

**Reply by glwagner** [reply]
Ok, here's what I understand. You have flow fields / velocity components on two domains: one domain covers a wide area and has a low resolution, and another velocity field covers a smaller domain within the larger domain, and has a higher resolution and perhaps more fidelity. So you'd like to advect Lagrangian particles over the large domain, but use the high-resolution velocities when the particles are within the limited area covered by the high resolution data. Is that right?

**Reply by glwagner** [reply]
If so Oceananigans does not have the capability to perform a single simulations on such a multi-resolution grid (we support stretched grids, but not unstructured grids). We also have not implemented the capability to perform two simulations that are coupled together, with one nested inside the other --- though it is in principle possible to construct such a coupled setup yourself using two different simulations that pass information between one another via `Callback`.

A simpler solution may be to define a single high resolution grid, and then to form a consistent interpolation that blends the low and high resolution data onto that one grid. Simulating particles with that interpolated flow field should possible. Moreover because particle advection uses linear interpolation anyways, I wonder if the result would be mathematically equivalent to using a nested setup as I described.

(Let me know if I'm interpreting this incorrectly...)

**Reply by Chinenyeani1986** [reply]
Yes, you are right. We want to advect particles over the smaller domain but use the low-resolution velocities when the particles are outside the smaller domain.

**Reply by glwagner** [reply]
Got it!

There's no pre-built utility to perform this kind of simulation out of the box. Here are the two possibilities I see:

1. Blend the two flow fields into a single field on one grid. I believe linear interpolation of the low resolution field onto an "expanded" high resolution field is what you want. If the fields are already identical on their boundaries then you are set; otherwise you may have to blend the fields within an overlap region, somehow.

2. Construct a framework in which there are two different particle advection simulations being performed at the same time. Next, implement boundary conditions via `Callback` (or, such a feature could be added to Oceananigans' source...) that pass particles from one domain to another.

Option 2 is basically developing the feature you're asking for in Oceananigans. It might not be a small amount of work, depending on your appetite for software development.

Option 1 seems more straightforward to me. In fact, this approach seems preferable to nested simulations anyways (it would make everything easier, including analysis and visualization). Another nice aspect of this approach is that because particle advection schemes typically use linear interpolation (which are composable?) of the velocity field, option 1 may be mathematically identical to a nested simulation (this should be verified). Thus option 1 might be preferred over option 2, in any code (even Ocean parcels). But perhaps I am missing something?

If Option 1 is not viable, we can open the implementation of nested simulations for particle advection as an issue and maybe it will get picked up at some point.
